File: test/json_values_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:weathergpt_mobile/core/models/json_values.dart';

void main() {
  group('jsonNum', () {
    test('passes through finite numbers, including zero and negatives', () {
      expect(jsonNum(0), 0);
      expect(jsonNum(-3.5), -3.5);
      expect(jsonNum(12), 12);
    });

    test('never turns a missing or unusable value into zero', () {
      expect(jsonNum(null), isNull);
      expect(jsonNum(''), isNull);
      expect(jsonNum('n/a'), isNull);
      expect(jsonNum([]), isNull);
      expect(jsonNum(double.nan), isNull);
      expect(jsonNum(double.infinity), isNull);
    });

    test('accepts numeric strings from loosely typed backends', () {
      expect(jsonNum(' 42 '), 42);
      expect(jsonNum('4.5'), 4.5);
    });
  });

  group('jsonInt', () {
    test('rejects genuinely fractional values instead of rounding silently',
        () {
      expect(jsonInt(3), 3);
      expect(jsonInt(3.0), 3);
      expect(jsonInt(3.4), isNull);
      expect(jsonInt(null), isNull);
    });
  });

  group('jsonString', () {
    test('treats serializer placeholders for absent values as null', () {
      expect(jsonString(null), isNull);
      expect(jsonString(''), isNull);
      expect(jsonString('   '), isNull);
      expect(jsonString('null'), isNull);
      expect(jsonString('None'), isNull);
      expect(jsonString('NaN'), isNull);
    });

    test('trims real values', () {
      expect(jsonString('  Partly cloudy '), 'Partly cloudy');
    });
  });

  group('jsonUtc', () {
    test('preserves the instant from an explicit offset', () {
      expect(jsonUtc('2026-09-19T14:30:00Z'),
          DateTime.utc(2026, 9, 19, 14, 30));
      // +05:30 must be converted, not read as a local wall clock.
      expect(jsonUtc('2026-09-19T20:00:00+05:30'),
          DateTime.utc(2026, 9, 19, 14, 30));
    });

    test('reads a naive timestamp as UTC and reports the assumption', () {
      expect(jsonUtc('2026-09-19T14:30:00'), DateTime.utc(2026, 9, 19, 14, 30));
      expect(naiveTimestampAssumedUtc('2026-09-19T14:30:00'), isTrue);
      expect(naiveTimestampAssumedUtc('2026-09-19T14:30:00Z'), isFalse);
      expect(naiveTimestampAssumedUtc('2026-09-19T20:00:00+05:30'), isFalse);
      expect(naiveTimestampAssumedUtc(null), isFalse);
    });

    test('malformed timestamps become null rather than throwing', () {
      expect(jsonUtc('not-a-date'), isNull);
      expect(jsonUtc(null), isNull);
    });
  });

  group('decodeJsonObject', () {
    test('tolerates empty and malformed bodies', () {
      expect(decodeJsonObject(null), isEmpty);
      expect(decodeJsonObject(''), isEmpty);
      expect(decodeJsonObject('{truncated'), isEmpty);
      expect(decodeJsonObject('[1,2]'), isEmpty);
      expect(decodeJsonObject('{"a":1}'), {'a': 1});
    });
  });
}

File: test/jsonValues.test.ts
import { describe, expect, it } from "vitest";

import { jsonBool, jsonDate, jsonDouble, jsonInt, jsonList, jsonMap, jsonNum, jsonString, naiveTimestampAssumedUtc, decodeJsonObject } from "../src/core/models/jsonValues";

describe("jsonNum", () => {
  it("passes finite numbers through", () => {
    expect(jsonNum(3.4)).toBe(3.4);
    expect(jsonNum(0)).toBe(0);
  });
  it("rejects NaN and Infinity", () => {
    expect(jsonNum(Number.NaN)).toBeNull();
    expect(jsonNum(Number.POSITIVE_INFINITY)).toBeNull();
  });
  it("parses numeric strings", () => {
    expect(jsonNum(" 12.5 ")).toBe(12.5);
    expect(jsonNum("abc")).toBeNull();
    expect(jsonNum("")).toBeNull();
  });
});

describe("jsonInt", () => {
  it("accepts integral values", () => {
    expect(jsonInt(4)).toBe(4);
    expect(jsonInt(4.0)).toBe(4);
    expect(jsonInt("7")).toBe(7);
  });
  it("rejects fractional values", () => {
    expect(jsonInt(4.5)).toBeNull();
  });
});

describe("jsonString", () => {
  it("trims and passes real strings", () => {
    expect(jsonString(" hi ")).toBe("hi");
  });
  it("nulls absent markers", () => {
    expect(jsonString("null")).toBeNull();
    expect(jsonString("None")).toBeNull();
    expect(jsonString("NaN")).toBeNull();
    expect(jsonString("")).toBeNull();
    expect(jsonString(undefined)).toBeNull();
  });
});

describe("jsonBool", () => {
  it("accepts booleans and true/false strings", () => {
    expect(jsonBool(true)).toBe(true);
    expect(jsonBool("false")).toBe(false);
    expect(jsonBool("yes")).toBeNull();
    expect(jsonBool(1)).toBeNull();
  });
});

describe("jsonMap / jsonList", () => {
  it("maps objects and nulls arrays", () => {
    expect(jsonMap({ a: 1 })).toEqual({ a: 1 });
    expect(jsonMap([1])).toBeNull();
    expect(jsonList([1, 2])).toEqual([1, 2]);
    expect(jsonList("nope")).toEqual([]);
  });
});

describe("jsonDate", () => {
  it("honours explicit offsets", () => {
    const d = jsonDate("2026-09-19T09:00:00+05:30");
    expect(d).not.toBeNull();
    expect(d!.getUTCHours()).toBe(3);
    expect(d!.getUTCMinutes()).toBe(30);
  });
  it("reads naive strings as UTC", () => {
    const d = jsonDate("2026-09-19T09:00");
    expect(d).not.toBeNull();
    expect(d!.getUTCHours()).toBe(9);
    expect(naiveTimestampAssumedUtc("2026-09-19T09:00")).toBe(true);
    expect(naiveTimestampAssumedUtc("2026-09-19T09:00Z")).toBe(false);
  });
  it("rejects garbage", () => {
    expect(jsonDate("not a date")).toBeNull();
    expect(jsonDate(null)).toBeNull();
  });
});

describe("decodeJsonObject", () => {
  it("parses objects", () => {
    expect(decodeJsonObject('{"a":1}')).toEqual({ a: 1 });
  });
  it("survives truncated bodies", () => {
    expect(decodeJsonObject('{"a":')).toEqual({});
    expect(decodeJsonObject(null)).toEqual({});
  });
});
